Add CLI for file entry add and resolve

nolash opened the pull request:
This PR adds a `bee-file` binary that creates file entries and also writes files to fs based on the metadata in the file entry.

Furthermore, it makes the following changes:

* add a chunk equality method
* add loggers to all non-bee cli commands
* no http default for split, now just returns swarmhash by default
* fix possible oversight in bee binary where logging logs to stdout

acud commented on cmd/bee-file/main.go:
```diff
@@ -0,0 +1,255 @@
+	}
+	defer outFile.Close()
+
+	err = cmdfile.JoinReadAll(j, e.Reference(), outFile)
```
can probably just return here `return cmdfile.Join....`

acud commented on cmd/bee-file/main.go:
```diff
@@ -0,0 +1,255 @@
+
+	// set up splitter to process the metadata
+	s := splitter.NewSimpleSplitter(stores)
+	ctx, cancel := context.WithCancel(context.Background())
```
i wonder if we need the cancel func at all. i cant see this being used except in the defer.

acud commented on cmd/bee-file/main.go:
```diff
@@ -0,0 +1,255 @@
+	}
+
+	// output reference to file entry
+	fmt.Println(fileEntryAddr)
```
i think the cobra `cmd` has some Println functionality that might be better

nolash replied:
"better" :D

acud replied:
:D

I think that its "better" because if you use the standard `fmt.Println` and the process quits, there's that strange `%` sign showing up when the process exits. I never understood why. The cobra print should mitigate that

acud commented on cmd/bee-join/main.go:
```diff
@@ -26,50 +22,18 @@ var (
+	logger, err = cmdfile.SetLogger(cmd, verbosity)
 	if err != nil {
-		return nil, err
+		fmt.Fprintln(os.Stderr, err.Error())
```
wouldn't returning the error from the function will result in this anyway?

acud commented on cmd/bee-split/main.go:
```diff
@@ -28,125 +23,23 @@ var (
-	var infile io.ReadCloser
+	logger, err = cmdfile.SetLogger(cmd, verbosity)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err.Error())
```
again, probably cleaner to return the error here no?

acud commented on cmd/bee/cmd/start.go:
```diff
@@ -56,15 +56,15 @@ func (c *command) initStartCmd() (err error) {
 				logger = logging.New(ioutil.Discard, 0)
 			case "1", "error":
-				logger = logging.New(cmd.OutOrStdout(), logrus.ErrorLevel)
+				logger = logging.New(cmd.OutOrStderr(), logrus.ErrorLevel)
```
why? i personally prefer to stdout, errors are anyway written to stderr afaik

nolash replied:
that's perfectly normal operation, to want the result of cli cmd and the log chatter kept separate.

janos replied:
Yes, log messages should be sent to stderr, as usual practice, as to have the command output to stdout as the result of the execution. I have set stdout here as start does not have the resulted output, only logs, but I have nothing against using stderr for log messages.

I would also like to hear from @svetomir or @vandot if this is ok from the infrastructure point of view.

acud replied:
all messages? error I understand (and i think the logger should do this regardless), but everything else? i am used that `WARN` and `ERROR` go to `stderr`, but the rest to `stdout` no?

nolash replied:
From a cli user point of view I would absolutely _loathe_ that, and I dare ever make the assumption that that's not the way it usually works.

janos replied:
This is true for applications that produce some output. In this case, `bee start` does not produce any output and it is a long running process, with logs that can be considered as the output. Maybe even this PR is not the place to change this. This change in `bee start` should be in a separate PR.

acud commented on cmd/file/io.go:
```diff
@@ -0,0 +1,235 @@
+}
+
+// Add adds a storage.Putter.
+func (t *TeeStore) Add(putter storage.Putter) {
```
why not in the constructor?
```
NewTeeStore(putters ...storage.Putter) *TeeStore {

}
```

Also, `NewTeeStore` should probably return a `storage.Putter`. No need to return the concrete type here. then `TeeStore` type can be unexported

nolash replied:
I find this easier to read

nolash replied:
> Also, NewTeeStore should probably return a storage.Putter.

:+1:

nolash replied:
Actually, if you look at `cmd/bee-split/main.go` you can see why having `Add` would be the better interface here.

acud replied:
well you could have alternatively:
```go
var stores []storage.Storer
if file { stores = append(stores, NewFileStore()) }
if http { stores = append(stores, NewHttpStore()) }
tee := NewTeeStore(stores...)
```
but this is a terribly tyrannic and selective-vision nitpick :)

nolash replied:
Aswell as ugly as f*****ck ;)

acud commented on cmd/file/io.go:
```diff
@@ -0,0 +1,235 @@
+			return nil, err
+		}
+	}
+	return nil, nil
```
note you're silently breaking the contract of the function signature. this is not so great as the caller might expect a slice here, resulting in a panic. i would expect at least to return a slice in the same size as the length of the variadic function parameter, all with `false` values

acud commented on cmd/file/io.go:
```diff
@@ -0,0 +1,235 @@
+			return nil, err
+		}
+	}
+	return nil, nil
```
here too

acud commented on cmd/file/io.go:
```diff
@@ -0,0 +1,235 @@
+			return nil, fmt.Errorf("upload failed: %v", res.Status)
+		}
+	}
+	return nil, nil
```
...

acud commented on cmd/file/io.go:
```diff
@@ -0,0 +1,235 @@
+		return 0, errors.New("overflow")
+	}
+	c, err := l.WriteCloser.Write(b)
+	l.total += int64(c)
```
if err is true, i'm not sure `total` should be incremented, or?

nolash replied:
from `go doc -all io | grep "type Writer interface -A10`:

```
Write writes len(p) bytes from p to the underlying data stream. It returns
    the number of bytes written from p (0 <= n <= len(p)) and any error
    encountered that caused the write to stop early.
```

in other words, `c` may be > 0 even if `err` is not `nil`.

acud commented on cmd/file/io.go:
```diff
@@ -0,0 +1,235 @@
+	return c, err
+}
+
+// Join reads all output from the provided joiner.
```
`JoinReadAll ....`

acud commented on cmd/file/io.go:
```diff
@@ -0,0 +1,235 @@
+	if total != l {
+		return fmt.Errorf("received only %d of %d total bytes", total, l)
+	}
+	return err
```
`return nil`?

acud commented on cmd/file/io.go:
```diff
@@ -0,0 +1,235 @@
+	case "0", "silent":
+		logger = logging.New(ioutil.Discard, 0)
+	case "1", "error":
+		logger = logging.New(cmd.OutOrStderr(), logrus.ErrorLevel)
```
not sure about stderr as the default for everything in this block

nolash replied:
It would be a very strange thing to log to stdout at all, from a cli usage perspective.

janos commented on cmd/file/io.go:
```diff
@@ -0,0 +1,238 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
```
Package `cmd/file` is not the `main` package and from its path it suggests that. I think that it would be more correct to utilize here internal package feature and the move this package to `cmd/internal/file`.

nolash replied:
Will it then still be accessible by `cmd/bee-split/main.go`

I was thinking of this structure (but left it like this now to make the diffs for split and join clearer)

```
cmd/file/io.go
cmd/file/bee-split/main.go
cmd/file/bee-join/main.go
cmd/file/bee-file/main.go
```

That still doesn't fit what you are describing as a problem though.

janos replied:
I would preserve that all first subdirectories under cmd are only main packages.

nolash replied:
I will try with the `internal` but I was under the impression that `internal` will only work for the package in its immediate parent directory...? in that case sharing code between those binaries is not possible using that method.

janos replied:
They just need to share the a common ancestor, which is here `github.com/ethersphere/bee/cmd`.

nolash commented:
@janos I've changed the dir to `internal`. And I have also reverted the `logging` change in `bee`. That was what we needed reply from @vandot @svetomir for right?

janos commented on cmd/bee-join/main.go:
```diff
@@ -26,50 +22,17 @@ var (
+	logger, err = cmdfile.SetLogger(cmd, verbosity)
 	if err != nil {
-		return nil, err
+		os.Exit(1)
```
We should log the error here. Or just return it as it was before?

janos commented on cmd/bee-split/main.go:
```diff
@@ -28,125 +23,22 @@ var (
-	var infile io.ReadCloser
+	logger, err = cmdfile.SetLogger(cmd, verbosity)
+	if err != nil {
+		os.Exit(1)
```
Also here, we can return the error.
